packages/backend/src/controllers/authentication.ts

```diff
@@ -73,6 +75,104 @@ export const getOpenIdUserFromRequest = ({
     return undefined;
 };
 
+export async function setupClient(): Promise<Client> {
+    const oktaIssuer = await Issuer.discover(
+        `https://${lightdashConfig.auth.okta.oktaDomain}/oauth2/default`,
+    );
+
+    const redirectUri = new URL(
+        `/api/v1${lightdashConfig.auth.okta.callbackPath}`,
+        lightdashConfig.siteUrl,
+    ).href;
+
+    const client = new oktaIssuer.Client({
+        client_id: lightdashConfig.auth.okta.oauth2ClientId ?? '',
+        client_secret: lightdashConfig.auth.okta.oauth2ClientSecret,
+        redirect_uris: [redirectUri],
+        response_types: ['code'],
+    });
+
+    return client;
+}
+
+export class OpenIDClientStrategy extends Strategy {
+    private clientPromise: Promise<Client>;
+
+    constructor() {
+        super();
+        this.clientPromise = setupClient();
+    }
+
+    async authenticate(req: Request) {
+        try {
+            const client = await this.clientPromise;
+
+            const redirectUri = new URL(
+                `/api/v1${lightdashConfig.auth.okta.callbackPath}`,
+                lightdashConfig.siteUrl,
+            ).href;
+
+            const params = client.callbackParams(req);
+            const tokenSet = await client.callback(redirectUri, params, {
+                code_verifier: req.session.oauth?.codeVerifier,
+                state: req.session.oauth?.state,
+            });
+
+            const userInfo = tokenSet.access_token
+                ? await client.userinfo(tokenSet.access_token)
+                : undefined;
+
+            try {
+                const { inviteCode } = req.session.oauth || {};
+                req.session.oauth = {};
+
+                if (!userInfo) return this.fail(401);
+
+                const { email } = userInfo;
+                const { sub: subject } = userInfo;
+                if (!(email && subject)) {
+                    return this.fail(
+                        {
+                            message: 'Could not parse authentication token',
+                        },
+                        401,
+                    );
+                }
+                const [fallbackFirstName, fallbackLastName] = (
+                    userInfo.name || ''
+                ).split(' ');
+
+                const openIdUser: OpenIdUser = {
+                    openId: {
+                        issuer: tokenSet.claims().iss,
+                        email,
+                        subject,
+                        firstName: userInfo.given_name || fallbackFirstName,
+                        lastName: userInfo.family_name || fallbackLastName,
+                        issuerType: OpenIdIdentityIssuerType.OKTA,
+                        groups: userInfo.groups as string[] | undefined,
+                    },
+                };
+                const user = await userService.loginWithOpenId(
+                    openIdUser,
+                    req.user,
+                    inviteCode,
+                );
```

Comment:
create resuable function - merge this with `genericOidcHandler`

Reply:
let's refactor this later, we can move the other services off passport generic handler